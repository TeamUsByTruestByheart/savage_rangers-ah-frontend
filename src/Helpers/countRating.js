export default (ratings) => {
  const x = ratings.map(item => ({
    first: item.rating * item.users,
    second: item.users
  }));
  return (
    x.map(item => item.first).reduce((a, b) => a + b, 0)
    / x.map(item => item.second).reduce((a, b) => a + b, 0)
  );
};
