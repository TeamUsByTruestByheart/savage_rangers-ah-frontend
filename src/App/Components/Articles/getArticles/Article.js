import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import propTypes from 'prop-types';
import Rater from 'react-rater';
import ReactHtmlParser from 'react-html-parser';
import ReactImageFallback from 'react-image-fallback';
import { Link } from 'react-router-dom';
import countRating from '../../../../Helpers/countRating';

export class Article extends Component {
  render() {
    const { data } = this.props;
    const {
      body, title, coverImage, createdAt, slug
    } = data;
    const { name } = data.Category;
    const { profileImage, firstName, username } = data.User;
    const { comments, reads, shares } = data.stats;
    const { allUsers, statistics } = data.rating;
    const ratings = countRating(statistics);

    const date = moment(createdAt).format('MMM Do YY');
    const img = 'https://ielektro.es/wp-content/uploads/2017/04/ventajas-comprar-LED.jpg';
    const defaultProfileImage = 'https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_960_720.png';
    const articleBody = body.length > 200 ? `${body.substring(0, 210)} ...` : body;
    const articleTitle = title.length > 52 ? `${title.substring(0, 52)}...` : title;
    const categoryColor = (name === 'LOVE' && '#ff1744') || (name === 'TECHNOLOGY' && '#3f51b5') || (name === 'MUSIC' && '#880e4f') || (name === 'BUSINESS' && '#2196f3') || (name === 'ART' && '#9c27b0') || '#afad99';
    const names = (firstName === null && username)
      || (firstName === undefined && username)
      || firstName;
    return (
      <React.Fragment>
        <div className="articles container row">
          <div className="articles__cover col-lg-3 col-md-12 col-sm-12">
            <ReactImageFallback
              src={coverImage}
              fallbackImage={img}
              initialImage="loader.gif"
              alt="cool image should be here"
              className="coverImage"
            />
          </div>
          <div className="articles__details col-lg-9 col-md-12 col-sm-12">
            <div className="articles__details--content">
              <Link className="link" to={`/articles/${slug}`}>
                <h3>
                  {articleTitle}
                </h3>
              </Link>
              <div className="category row">
                <div className="col-lg-6 col-md-6 col-sm-12 row">
                  <div className="color" style={{ background: `${categoryColor}` }} />
                  <div className="name col ">
                    <h5>{name}</h5>
                  </div>

                </div>
                <div className="Rate col-lg-6 col-md-6 col-sm-12 row">
                  <div className="rating col">
                    <Rater total={5} rating={ratings} interactive={false} />
                  </div>
                  <div className="ratedUser col-1">
                    <h6>
                      [
                      {allUsers}
                      ]
                    </h6>
                  </div>

                </div>
              </div>
              <p className="articleBody">
                {ReactHtmlParser(articleBody)}
              </p>
            </div>
            <div className="articles__details--Author row">
              <div className="col-lg-2 col-md-6 col-sm-6">
                <ReactImageFallback
                  src={profileImage}
                  fallbackImage={defaultProfileImage}
                  initialImage="loader.gif"
                  alt="cool image should be here"
                  className="proImage"
                />
              </div>
              <div className="profileDetail col-lg-3 col-md-6 col-sm-6">
                <Link className="userProfile link" to={`/profile/${username}`}>
                  <p className="pro">
                    {names}
                    {' '}
                    {data.User.lastName}
                  </p>
                </Link>
                <p>{date}</p>
              </div>
              <div className="articlesBtn col-lg-6 col-md-6 col-sm-12">
                <div className="row">
                  <div className="btn-articles">
                    <button type="button" className="btn-article fa">
                      <i className="fab fa-readme" />
                      {' '}
                      {''}
                      {reads}
                    </button>
                  </div>
                  <div className="btn-articles">
                    <button type="button" className="btn-article fa">
                      <i className="far fa-comment-alt" />
                      {' '}
                      {''}
                      {comments}
                    </button>
                  </div>
                  <div className="btn-articles">
                    <button type="button" className="btn-article fa">
                      <i className="fas fa-share-alt" />
                      {' '}
                      {''}
                      {shares}
                    </button>
                  </div>
                </div>

                <div className="more">
                  <Link className="read-more fa link" to={`/articles/${slug}`}><p>Read more</p></Link>
                </div>
              </div>

            </div>
          </div>

        </div>
      </React.Fragment>
    );
  }
}

Article.propTypes = { data: propTypes.object };

export default connect(state => ({ stats: state.articleData }))(Article);
