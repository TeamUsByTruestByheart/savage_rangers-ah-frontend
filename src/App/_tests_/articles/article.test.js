import React from 'react';
import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import configureStore from 'redux-mock-store';
import { shallow } from '../../../enzyme';
import Article from '../../Components/Articles/getArticles/Article';
import countRating from '../../../Helpers/countRating';

const middlewares = [thunk, promiseMiddleware];
const mockStore = configureStore(middlewares);

describe('<articles />', () => {
  describe('render()', () => {
    it('should give rating', () => {
      const statistics = [
        { rating: 1, users: 1, percentage: 50 },
        { rating: 5, users: 1, percentage: 50 }
      ];
      expect(countRating(statistics)).toBe(3);
    });

    it('it should render a single article', () => {
      const get = jest.fn(() => { });
      const props = {
        getArticles: get,
        article: { articles: [{}] },
        data: {
          body: 'hello',
          title: 'hi',
          coverImage: 'image',
          createdAt: '2019-08-14T12:48:02',
          Category: { name: 'LOVE' },
          User: {
            firstName: 'Alain',
            lastName: 'Burindi',
            profileImage: 'noimage.jpg',
            username: 'Burindi'
          },
          rating: { allUsers: 0, statistics: Array(0) },
          stats: { reads: 9, shares: 5, comments: 4 }

        }
      };

      const normalStats = { testRedux: {} };
      const store = mockStore(normalStats);
      const article = shallow(<Article store={store} {...props} />);
      expect(Article).toMatchSnapshot();
      expect(article.find('ContextProvider').exists()).toBe(true);
      expect(article.find('Article').exists()).toBe(true);
      article.render();
    });

    it('it should render a single with deferent data', () => {
      const get = jest.fn(() => { });
      const props = {
        getArticles: get,
        article: { articles: [{}] },
        data: {
          body: 'Flash-forward 100 years, and mental health experts today echo many of Sadler’s sentiments — albeit using different language. “Worry is part of human nature,” says Robert Leahy, a New York-based clinical psychologist and associate editor of the International Journal of Cognitive Therapy',
          title: 'Most Things You Worry About Will Never Actually Happen',
          coverImage: 'default.jpeg',
          createdAt: '2019-08-14T12:48:02',
          Category: { name: 'TECHNOLOGY' },
          User: {
            firstName: null,
            lastName: null,
            profileImage: 'defaultAvatar.jpg',
            username: 'Burindi'
          },
          rating: { allUsers: 0, statistics: Array(0) },
          stats: { reads: 9, shares: 5, comments: 4 }

        }
      };

      const normalStats = { testRedux: {} };
      const store = mockStore(normalStats);
      const article = shallow(<Article store={store} {...props} />);
      expect(Article).toMatchSnapshot();
      expect(article.find('ContextProvider').exists()).toBe(true);
      expect(article.find('Article').exists()).toBe(true);
      article.render();
    });

    it('it should render a single with deferent data to cover all lines', () => {
      const get = jest.fn(() => { });
      const props = {
        getArticles: get,
        article: { articles: [{}] },
        data: {
          body: 'hello',
          title: 'Most Things You Worry About Will Never Actually Happen',
          coverImage: null,
          createdAt: '2019-08-14T12:48:02',
          Category: { name: 'MUSIC' },
          User: {
            firstName: null,
            lastName: null,
            profileImage: null,
            username: 'Burindi'
          },
          rating: { allUsers: 0, statistics: Array(0) },
          stats: { reads: 9, shares: 5, comments: 4 }

        }
      };

      const normalStats = { testRedux: {} };
      const store = mockStore(normalStats);
      const article = shallow(<Article store={store} {...props} />);
      expect(Article).toMatchSnapshot();
      expect(article.find('ContextProvider').exists()).toBe(true);
      expect(article.find('Article').exists()).toBe(true);
      article.render();
    });

    it('it should render a single with deferent data to cover all lines', () => {
      const get = jest.fn(() => { });
      const props = {
        getArticles: get,
        article: { articles: [{}] },
        data: {
          body: 'hello',
          title: 'Most Things You Worry About Will Never Actually Happen',
          coverImage: null,
          createdAt: '2019-08-14T12:48:02',
          Category: { name: 'BUSINESS' },
          User: {
            firstName: null,
            lastName: null,
            profileImage: 'holler',
            username: 'Burindi'
          },
          rating: { allUsers: 0, statistics: Array(0) },
          stats: { reads: 9, shares: 5, comments: 4 }
        }
      };

      const normalStats = { testRedux: {} };
      const store = mockStore(normalStats);
      const article = shallow(<Article store={store} {...props} />);
      expect(Article).toMatchSnapshot();
      expect(article.find('ContextProvider').exists()).toBe(true);
      expect(article.find('Article').exists()).toBe(true);
      article.render();
    });

    it('it should render a single with deferent data to cover all lines', () => {
      const get = jest.fn(() => { });
      const props = {
        getArticles: get,
        article: { articles: [{}] },
        data: {
          body: 'hello',
          title: 'Most Things You Worry About Will Never Actually Happen',
          coverImage: null,
          createdAt: '2019-08-14T12:48:02',
          Category: { name: 'ART' },
          User: {
            firstName: null,
            lastName: null,
            profileImage: 'holler',
            username: 'Burindi'
          },
          rating: { allUsers: 0, statistics: Array(0) },
          stats: { reads: 9, shares: 5, comments: 4 }

        }
      };

      const normalStats = { testRedux: {} };
      const store = mockStore(normalStats);
      const article = shallow(<Article store={store} {...props} />);
      expect(Article).toMatchSnapshot();
      expect(article.find('ContextProvider').exists()).toBe(true);
      expect(article.find('Article').exists()).toBe(true);
      article.render();
    });

    it('it should render a single with deferent data to cover all lines', () => {
      const get = jest.fn(() => { });
      const props = {
        getArticles: get,
        article: { articles: [{}] },
        data: {
          body: 'hello',
          title: 'Most Things You Worry About Will Never Actually Happen',
          coverImage: null,
          createdAt: '2019-08-14T12:48:02',
          Category: { name: 'KAKAKA' },
          User: {
            firstName: null,
            lastName: null,
            profileImage: 'holler',
            username: 'Burindi'
          },
          rating: { allUsers: 0, statistics: Array(0) },
          stats: { reads: 9, shares: 5, comments: 4 }

        }
      };

      const normalStats = { testRedux: {} };
      const store = mockStore(normalStats);
      const article = shallow(<Article store={store} {...props} />);
      expect(Article).toMatchSnapshot();
      expect(article.find('ContextProvider').exists()).toBe(true);
      expect(article.find('Article').exists()).toBe(true);
      article.render();
    });
  });
});
