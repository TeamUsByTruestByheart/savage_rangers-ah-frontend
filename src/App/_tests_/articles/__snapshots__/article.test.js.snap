// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<articles /> render() it should render a single article 1`] = `
Object {
  "$$typeof": Symbol(react.memo),
  "WrappedComponent": [Function],
  "compare": null,
  "displayName": "Connect(Article)",
  "type": [Function],
}
`;

exports[`<articles /> render() it should render a single with deferent data 1`] = `
Object {
  "$$typeof": Symbol(react.memo),
  "WrappedComponent": [Function],
  "compare": null,
  "displayName": "Connect(Article)",
  "type": [Function],
}
`;

exports[`<articles /> render() it should render a single with deferent data to cover all lines 1`] = `
Object {
  "$$typeof": Symbol(react.memo),
  "WrappedComponent": [Function],
  "compare": null,
  "displayName": "Connect(Article)",
  "type": [Function],
}
`;

exports[`<articles /> render() it should render a single with deferent data to cover all lines 2`] = `
Object {
  "$$typeof": Symbol(react.memo),
  "WrappedComponent": [Function],
  "compare": null,
  "displayName": "Connect(Article)",
  "type": [Function],
}
`;

exports[`<articles /> render() it should render a single with deferent data to cover all lines 3`] = `
Object {
  "$$typeof": Symbol(react.memo),
  "WrappedComponent": [Function],
  "compare": null,
  "displayName": "Connect(Article)",
  "type": [Function],
}
`;

exports[`<articles /> render() it should render a single with deferent data to cover all lines 4`] = `
Object {
  "$$typeof": Symbol(react.memo),
  "WrappedComponent": [Function],
  "compare": null,
  "displayName": "Connect(Article)",
  "type": [Function],
}
`;
